File: benchmarks/pandas/bench_mode_dataframe_fn.py
"""
Benchmark: pandas DataFrame.mode() — column-wise mode.
Outputs JSON: {"function": "mode_dataframe_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

ROWS = 10_000
WARMUP = 5
ITERATIONS = 20

df = pd.DataFrame({
    "a": [i % 10 for i in range(ROWS)],
    "b": [float("nan") if i % 50 == 0 else i % 5 for i in range(ROWS)],
    "c": [i % 3 for i in range(ROWS)],
})

for _ in range(WARMUP):
    df.mode()
    df.mode(dropna=False)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.mode()
    df.mode(dropna=False)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "mode_dataframe_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_mode_dataframe_fn.ts
/**
 * Benchmark: modeDataFrame — standalone functional mode for DataFrame columns.
 * Outputs JSON: {"function": "mode_dataframe_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, modeDataFrame } from "../../src/index.ts";

const ROWS = 10_000;
const WARMUP = 5;
const ITERATIONS = 20;

// Low-cardinality numeric data (many ties → large mode arrays)
const df = DataFrame.fromColumns({
  a: Array.from({ length: ROWS }, (_, i) => i % 10),
  b: Array.from({ length: ROWS }, (_, i) => (i % 50 === 0 ? null : i % 5)),
  c: Array.from({ length: ROWS }, (_, i) => (i % 3)),
});

for (let i = 0; i < WARMUP; i++) {
  modeDataFrame(df);
  modeDataFrame(df, { dropna: false });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  modeDataFrame(df);
  modeDataFrame(df, { dropna: false });
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "mode_dataframe_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
